Lib/ufo2ft/featureWriters/gdefFeatureWriter.py

```diff
@@ -14,34 +12,41 @@ class GdefFeatureWriter(BaseFeatureWriter):
     """
 
     tableTag = "GDEF"
+    features = frozenset(["GlyphClassDefs", "LigatureCarets"])
+    insertFeatureMarker = None
 
     def setContext(self, font, feaFile, compiler=None):
-        ctx = self.context = SimpleNamespace(
-            font=font,
-            feaFile=feaFile,
-            compiler=compiler,
-            todo=set(),
-        )
-        # skip if a GDEF is in the features
-        if ast.findTable(self.context.feaFile, "GDEF") is not None:
-            return ctx
+        ctx = super().setContext(font, feaFile, compiler=compiler)
+
+        ctx.gdefTableBlock = ast.findTable(self.context.feaFile, "GDEF")
+        if ctx.gdefTableBlock:
+            for fea in ctx.gdefTableBlock.statements:
+                if isinstance(fea, ast.GlyphClassDefStatement):
+                    ctx.todo.remove("GlyphClassDefs")
+                elif isinstance(fea, ast.LigatureCaretByIndexStatement) or isinstance(
+                    fea, ast.LigatureCaretByPosStatement
+                ):
+                    ctx.todo.remove("LigatureCarets")
```

Comment:
not sure if techically possible to have multiple such statements in a GDEF block, but would that `.remove()` call raise an error if you attempt to remove it more than once? Also maybe you can break if todo is emptied

Reply:
We can use `discard()` to remove it if present.